contrib/win32/win32compat/misc.c

```diff
@@ -1115,216 +1098,192 @@ get_machine_domain_name(wchar_t *domain, int size)
+		goto cleanup;
+	}
 
-/* This method will return in "group@domain" format */
-char *
-append_domain_to_groupname(char *groupname)
-{
-	if(!groupname) return NULL;
+	/* read group sids from logon token */
+	DWORD group_size = 0;
+	if (GetTokenInformation(logon_token, TokenGroups, NULL, 0, &group_size) == 0 && GetLastError() != ERROR_INSUFFICIENT_BUFFER ||
+		GetTokenInformation(logon_token, TokenGroups, (PTOKEN_GROUPS)(group_buf =
+		(PTOKEN_GROUPS)malloc(group_size)), group_size, &group_size) == 0) {
+		goto cleanup;
+	}
 
-	int len = (int) strlen(machine_domain_name) + (int) strlen(groupname) + 2;
-	char *groupname_with_domain = malloc(len);
-	if(!groupname_with_domain) {
-		error("failed to allocate memory!");
-		return NULL;
+	/* open a new connection to the lsa policy provider for group lookup */
+	LSA_OBJECT_ATTRIBUTES ObjectAttributes;
+	ZeroMemory(&ObjectAttributes, sizeof(ObjectAttributes));
+	if (LsaOpenPolicy(NULL, &ObjectAttributes, POLICY_LOOKUP_NAMES, &lsa_policy) != 0) {
+		goto cleanup;
 	}
 
-	strcpy_s(groupname_with_domain, len, groupname);
-	strcat_s(groupname_with_domain, len, "@");
-	strcat_s(groupname_with_domain, len, machine_domain_name);	
+	/* translate all the sids to real group names*/
+	group_sids = (PSID *)malloc(sizeof(PSID) * group_buf->GroupCount);
+	for (DWORD i = 0; i < group_buf->GroupCount; i++) group_sids[i] = group_buf->Groups[i].Sid;
+	if (LsaLookupSids(lsa_policy, group_buf->GroupCount, group_sids, &domain_list, &name_list) != 0) {
+		goto cleanup;
+	}
 
-	groupname_with_domain[len-1]= '\0';
+	/* enumerate all groups and add to group list */
+	user_groups = (char**)malloc(sizeof(char*) * group_buf->GroupCount);
+	for (DWORD group_index = 0; group_index < group_buf->GroupCount; group_index++) {
 
-	return groupname_with_domain;
-}
+		/* ignore groups that are not local or domain groups */
+		if ((group_buf->Groups[group_index].Attributes & SE_GROUP_ENABLED) == 0) continue;
+		if (name_list[group_index].Use != SidTypeGroup && name_list[group_index].Use != SidTypeAlias) continue;
 
-void
-populate_user_groups(char **group_name, int *group_index, DWORD groupsread, DWORD totalgroups, LPBYTE buf, group_type groupType)
-{
-	if(0 == groupsread) return;
-	char *user_group_name = NULL;
-		
-	if (groupType == GLOBAL_UNIVERSAL_GROUP) {
-		LPGROUP_USERS_INFO_0 pTmpBuf = (LPGROUP_USERS_INFO_0)buf;
-		for (DWORD i = 0; (i < groupsread) && pTmpBuf; i++, pTmpBuf++) {
-			if (!(user_group_name = utf16_to_utf8(pTmpBuf->grui0_name))) {
-				error("utf16_to_utf8 failed to convert:%ls", pTmpBuf->grui0_name);
-				return;
-			}
+		/* create the group list as groupname@netbios except for local group */
+		int current_group = (*ngroups)++;
+		PLSA_UNICODE_STRING domain = &(domain_list->Domains[name_list[group_index].DomainIndex].Name);
+		int name_size = snprintf(NULL, 0, "%wZ@%wZ", &(name_list[group_index].Name), domain) + 1;
+		sprintf_s((user_groups[current_group] = (char*)malloc(name_size)), name_size,
```

Comment:
this code assumes malloc will not fail.
if malloc fails then it will be difficult to debug so please have a malloc NULL check.